src/Environments/include/State.h

```diff
@@ -1,10 +1,32 @@
 #ifndef MCTS_LIBRARY_STATE_H
 #define MCTS_LIBRARY_STATE_H
 
-template <class T> class State {
-public:
-  State(T data) : m_data(data){};
-  T m_data;
+#include <any>
+#include <memory>
+
+struct Concept {
+    virtual ~Concept() {}
+    virtual std::any getData() const = 0;
+};
+
+template <typename T> struct Model : Concept {
+    Model(const T &t) : m_data(t) {}
+
+    std::any getData() const override { return m_data; }
+
+  private:
+    T m_data;
+};
+
+class State {
+  public:
+    template <typename T> State(T data) : m_data(std::make_unique<Model<T>>(std::move(data))){};
```

Comment:
Move destroys the original copy of data which may surprise the caller if data is not a reference type.
Because data is a template type, State doesn't know if this is the case, or if T even has a move constructor.
If this is intentional, please document

Reply:
If data is a reference type* obviously there's no problem if it's a value type

Reply:
True, will leave it out for now. But in the future we might consider to put it in, while finding some way to ensure it is used correctly. In some cases it might provide a noticeable performance boost if used correctly.

Reply:
I agree, but premature optimizations and all that. Regardless, i think we should profile the algorithm before we start optimizing